rust/src/lockfile.rs

```diff
@@ -18,23 +18,47 @@ use std::io;
 use crate::utils;
 
 /// Parse a JSON lockfile definition.
-fn lockfile_parse_stream<R: io::Read>(input: &mut R,) -> Fallible<LockfileConfig> {
-    let lockfile: LockfileConfig = serde_json::from_reader(input).map_err(|e| {
-        io::Error::new(
-            io::ErrorKind::InvalidInput,
-            format!("serde-json: {}", e.to_string()),
-        )
-    })?;
+fn lockfile_parse_stream<R: io::Read>(
+    fmt: utils::InputFormat,
+    input: &mut R,
+) -> Fallible<LockfileConfig> {
+    let lockfile: LockfileConfig = match fmt {
+        utils::InputFormat::JSON => {
+            let lf: LockfileConfig = serde_json::from_reader(input).map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("serde-json: {}", e.to_string()),
+                )
+            })?;
+            lf
+        }
+        utils::InputFormat::YAML => {
+            let lf: LockfileConfig = serde_yaml::from_reader(input).map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("serde-yaml: {}", e.to_string()),
+                )
+            })?;
+            lf
+        }
+    };
     Ok(lockfile)
 }
 
 /// Given a lockfile filename, parse it
 fn lockfile_parse<P: AsRef<Path>>(filename: P,) -> Fallible<LockfileConfig> {
     let filename = filename.as_ref();
+    let basename = filename
+        .file_name()
+        .map(|s| s.to_string_lossy())
+        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Expected a filename"))?;
+    let fmt = if basename.ends_with(".yaml") || basename.ends_with(".yml") {
```

Comment:
Would be cleaner as a `fn` associated with `InputFormat`, like:

```
fn detect_from_filename<P: AsRef<Path>>(name: P) -> Self {
    let name = name.as_ref();
    if name.ends_with(".yaml") || basename.ends_with(".yml") {
        Self::YAML
    } else {
        Self::JSON
    };
}
```

or so.

Reply:
Done.